crates/noirc_evaluator/src/ssa/ssa_gen/mod.rs

```diff
@@ -241,9 +255,27 @@ impl<'a> FunctionContext<'a> {
     }
 
     fn codegen_index(&mut self, index: &ast::Index) -> Values {
-        let array = self.codegen_non_tuple_expression(&index.collection);
+        let array_or_slice = self.codegen_expression(&index.collection).into_value_list(self);
         let index_value = self.codegen_non_tuple_expression(&index.index);
-        self.codegen_array_index(array, index_value, &index.element_type, index.location)
+        if array_or_slice.len() > 1 {
+            let slice_length = array_or_slice[0];
+            let slice = array_or_slice[1];
+            self.codegen_array_index(
+                slice,
+                index_value,
+                &index.element_type,
+                index.location,
+                Some(slice_length),
+            )
+        } else {
+            self.codegen_array_index(
+                array_or_slice[0],
+                index_value,
+                &index.element_type,
+                index.location,
+                None,
+            )
+        }
```

Comment:
```suggestion
        let (array, slice_length) = if array_or_slice.len() > 1 {
            (array_or_slice[1], Some(array_or_slice[0]))
        } else {
            (array_or_slice[0], None)
        };
        self.codegen_array_index(array, index_value, &index.element_type, index.location, slice_length)
```
Not sure if the last line will fit on one line. Maybe if `index.element_type` or `index.location` are also separate variables.